<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>complexity - quasitools</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  <link href="../stylesheets/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "complexity";
    var mkdocs_page_input_path = "complexity.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> quasitools</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../requirements/">Requirements</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../installation/">Installation</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../formats/">Data Formats</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Tools</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../aacoverage/">aacoverage</a>
                </li>
                <li class="">
                    
    <a class="" href="../aavar/">call aavr</a>
                </li>
                <li class="">
                    
    <a class="" href="../codonvar/">call codonvar</a>
                </li>
                <li class="">
                    
    <a class="" href="../ntvar/">call ntvar</a>
                </li>
                <li class=" current">
                    
    <a class="current" href="./">complexity</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#complexity">complexity</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#basic-usage">Basic Usage</a></li>
        
            <li><a class="toctree-l4" href="#arguments">Arguments</a></li>
        
            <li><a class="toctree-l4" href="#options">Options</a></li>
        
            <li><a class="toctree-l4" href="#output">Output</a></li>
        
            <li><a class="toctree-l4" href="#applications">Applications</a></li>
        
            <li><a class="toctree-l4" href="#example-fasta-reads-file">Example: FASTA Reads File</a></li>
        
            <li><a class="toctree-l4" href="#example-bam-file-with-a-reference-fasta">Example: BAM File With a Reference FASTA</a></li>
        
        </ul>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../consensus/">consensus</a>
                </li>
                <li class="">
                    
    <a class="" href="../distance/">distance</a>
                </li>
                <li class="">
                    
    <a class="" href="../dnds/">dnds</a>
                </li>
                <li class="">
                    
    <a class="" href="../drmutations/">drmutations</a>
                </li>
                <li class="">
                    
    <a class="" href="../hydra/">hydra</a>
                </li>
                <li class="">
                    
    <a class="" href="../quality/">quality</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../contact/">Contact</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../legal/">Legal</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">quasitools</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>Tools &raquo;</li>
        
      
    
    <li>complexity</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="complexity">complexity</h1>
<p>Reports various measures of viral quasispecies complexity.</p>
<h2 id="basic-usage">Basic Usage</h2>
<p>FASTA Input:</p>
<pre><code class="bash">quasitools complexity fasta [OPTIONS] &lt;FASTA READS FILE&gt;
</code></pre>

<p>BAM Input:</p>
<pre><code class="bash">quasitools complexity bam &lt;FASTA REFERENCE FILE&gt; &lt;BAM FILE&gt; &lt;K-MER SIZE&gt; [OPTIONS] 
</code></pre>

<h2 id="arguments">Arguments</h2>
<h3 id="fasta-reads-file">FASTA Reads File</h3>
<p>This input file is only necessary when running the tool in FASTA mode.</p>
<p>An aligned FASTA file containing multiple aligned sequences, representing haplotypes of a genomic region from the mutant spectrum of interest. This FASTA file would likely be created using a multiple sequence alignment tool from aligning amplicon sequencing data.</p>
<h3 id="fasta-reference-file">FASTA Reference File</h3>
<p>This input file is only necessary when running the tool in BAM mode.</p>
<p>A reference file of the sequence of interest. The BAM file must be generated using this reference file.</p>
<h3 id="bam-file">BAM File</h3>
<p>This input file is only necessary when running the tool in BAM mode.</p>
<p>A BAM file describing the alignments of reads to the same reference provided as input. These reads should be derived from a quasispecies mutant spectrum. This BAM file would likely be created using a read aligner which aligns FASTQ reads to a FASTA reference.</p>
<h3 id="k-mer-size">k-mer Size</h3>
<p>This input is only necessary when running the tool in BAM mode.</p>
<p>The <em>k</em>-mer size defines the length of the <em>k</em>-mer sequence fragments. A sliding window of this <em>k</em>-mer size is used to scan across the reference genome. The sequences at each sliding window are used to calculate the quasispecies complexity.</p>
<h2 id="options">Options</h2>
<h3 id="filter">FILTER</h3>
<pre><code>-f [INTEGER]
</code></pre>

<p>This option is only available when running the tool in BAM mode.</p>
<p>This option allows for a user-defined filter size between 0 and 100. Haplotypes under the filter size will not be used when calculating the quasispecies complexity at a particular position in the genome.</p>
<h3 id="output-file">OUTPUT FILE</h3>
<pre><code>-o [USER-DEFINED-FILE-NAME.CSV]
</code></pre>

<p>This option is availble when running the tool in both BAM and FASTA mode.</p>
<p>This option allows users to define an output file location, where the program output will be written in CSV format.</p>
<h2 id="output">Output</h2>
<p>The quasispecies complexity measures are taken from <em>Gregori, Josep, et al. 2016</em>. These include the following various indices:</p>
<p>Incidence (Entity Level):</p>
<ul>
<li>Number of haplotypes</li>
<li>Number of polymorphic sites</li>
<li>Number of mutations</li>
</ul>
<p>Abundance (Molecular Level):</p>
<ul>
<li>Shannon entropy</li>
<li>Simpson index</li>
<li>Gini-Simpson index</li>
<li>Hill numbers</li>
</ul>
<p>Functional (Incidence):</p>
<ul>
<li>Minimum mutation frequency (Mf min)</li>
<li>Mutation frequency (Mfe)</li>
<li>FAD</li>
<li>Sample nucleotide diversity</li>
</ul>
<p>Functional (Abundance):</p>
<ul>
<li>Maximum mutation frequency (Mf max)</li>
<li>Population nucleotide diversity</li>
</ul>
<h2 id="applications">Applications</h2>
<ul>
<li>Assessing the quasispecies complexity of a genomic region.</li>
<li>Comparing the quasispecies complexity of multiple genomic regions from the same mutant spectrum.</li>
</ul>
<h2 id="example-fasta-reads-file">Example: FASTA Reads File</h2>
<h3 id="data">Data</h3>
<p>The following example data may be used to run the tool:</p>
<ul>
<li><a href="../data/aligned.fasta">aligned.fasta</a></li>
</ul>
<h3 id="command">Command</h3>
<pre><code class="bash">quasitools complexity fasta aligned.fasta
</code></pre>

<h3 id="output_1">Output</h3>
<pre><code class="text">Position,Number of Haplotypes,Haplotype Population,Number of Polymorphic Sites,Number of Mutations,Shannon Entropy,Shannon Entropy Normalized to N,Shannon Entropy Normalized to H,Simpson Index,Gini-Simpson Index,Hill Number #0,HIll Number #1,Hill Number #2,Hill Number #3,Minimum Mutation Frequency,Mutation Frequency,Functional Attribute Diversity,Sample Nucleotide Diversity (Entity),Maximum Mutation Frequency,Population Nucleotide Diversity,Sample Nucleotide Diversity
0,9,30,38,40,1.8774672554524843,0.5520018525167073,0.8544721713101401,0.19111111111111112,0.8088888888888889,9.0,6.536927510444632,5.232558139534883,4.543368996115371,0.013333333333333334,0.05555555555555555,7.379999999999999,0.10249999999999998,0.03866666666666667,0.06682222222222223,0.06912643678160921
</code></pre>

<h2 id="example-bam-file-with-a-reference-fasta">Example: BAM File With a Reference FASTA</h2>
<h3 id="data_1">Data</h3>
<p>The following example data may be used to run the tool:</p>
<ul>
<li><a href="../data/generated.fasta">generated.fasta</a></li>
<li><a href="../data/generated.bam">generated.bam</a></li>
<li><a href="../data/generated.bai">generated.bai</a></li>
</ul>
<h3 id="command_1">Command</h3>
<pre><code class="bash">quasitools complexity bam generated.fasta generated.bam 200
</code></pre>

<h3 id="output_2">Output</h3>
<pre><code class="text">Position,Number of Haplotypes,Haplotype Population,Number of Polymorphic Sites,Number of Mutations,Shannon Entropy,Shannon Entropy Normalized to N,Shannon Entropy Normalized to H,Simpson Index,Gini-Simpson Index,Hill Number #0,HIll Number #1,Hill Number #2,Hill Number #3,Minimum Mutation Frequency,Mutation Frequency,Functional Attribute Diversity,Sample Nucleotide Diversity (Entity),Maximum Mutation Frequency,Population Nucleotide Diversity,Sample Nucleotide Diversity
0,6,6,6,15,1.7917594692280547,0.9999999999999999,0.9999999999999999,0.16666666666666669,0.8333333333333333,6.0,5.999999999999998,5.999999999999999,6.000000000000001,0.0125,0.01916666666666667,0.7300000000000004,0.02433333333333335,0.019166666666666665,0.020277777777777773,0.02433333333333333
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../consensus/" class="btn btn-neutral float-right" title="consensus">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../ntvar/" class="btn btn-neutral" title="call ntvar"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
      <p>Copyright Government of Canada 2017-2019</p>
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../ntvar/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../consensus/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
