<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>call aavr - quasitools</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  <link href="../stylesheets/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "call aavr";
    var mkdocs_page_input_path = "aavar.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> quasitools</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../requirements/">Requirements</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../installation/">Installation</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../formats/">Data Formats</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Tools</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../aacoverage/">aacoverage</a>
                </li>
                <li class=" current">
                    
    <a class="current" href="./">call aavr</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#aavar">aavar</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#basic-usage">Basic Usage</a></li>
        
            <li><a class="toctree-l4" href="#arguments">Arguments</a></li>
        
            <li><a class="toctree-l4" href="#options">Options</a></li>
        
            <li><a class="toctree-l4" href="#output_1">Output</a></li>
        
            <li><a class="toctree-l4" href="#examples">Examples</a></li>
        
        </ul>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../codonvar/">call codonvar</a>
                </li>
                <li class="">
                    
    <a class="" href="../ntvar/">call ntvar</a>
                </li>
                <li class="">
                    
    <a class="" href="../complexity/">complexity</a>
                </li>
                <li class="">
                    
    <a class="" href="../consensus/">consensus</a>
                </li>
                <li class="">
                    
    <a class="" href="../distance/">distance</a>
                </li>
                <li class="">
                    
    <a class="" href="../dnds/">dnds</a>
                </li>
                <li class="">
                    
    <a class="" href="../drmutations/">drmutations</a>
                </li>
                <li class="">
                    
    <a class="" href="../hydra/">hydra</a>
                </li>
                <li class="">
                    
    <a class="" href="../quality/">quality</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../contact/">Contact</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../legal/">Legal</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">quasitools</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>Tools &raquo;</li>
        
      
    
    <li>call aavr</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="aavar">aavar</h1>
<p>Call amino acid mutations from a <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">BAM</a> alignment file and a supplied reference file. Please refer to <a href="../formats">Data Formats</a> for detailed information about the the expected input formats for this tool.</p>
<h2 id="basic-usage">Basic Usage</h2>
<pre><code class="bash">quasitools aavar [options] &lt;BAM file&gt; &lt;reference file&gt; &lt;bed file&gt; [variants file] [mutation db]
</code></pre>

<h2 id="arguments">Arguments</h2>
<h3 id="bam-file">BAM File</h3>
<p>A <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">BAM</a> file (.bam) of sequences aligned to a related reference. A BAM index file (.bai) is also required and should be named the same as the BAM file, with the extension instead changed from ".bam" to ".bai".</p>
<h3 id="reference-file">Reference File</h3>
<p>A reference file related to the aligned sequences in the BAM file. The provided reference file must be the same reference file used when producing the BAM and BAM index files.</p>
<h3 id="bed-file">BED File</h3>
<p>A <a href="https://bedtools.readthedocs.io/en/latest/content/general-usage.html">BED file</a> that specifies the coordinates of genes, with repsect to the provided reference. This BED file must be a BED4+ file and therefore contain at least the first 4 BED file columns. The "names" of these genetic regions in the BED4 file must be the same names used in the "genetic regions" column of the mutation database. Please refer to <a href="../formats">Data Formats</a> for more information.</p>
<h3 id="variants-file">Variants File</h3>
<p>A <a href="https://en.wikipedia.org/wiki/Variant_Call_Format">VCF (Variant Call Format)</a> file format specifying identified variants in the BAM file, with respect to the passed reference file. When this file is provided, the computational running time of the program is improved. This variants file should be generated using the same BAM and reference files passed as parameters to program. The VCF output of <a href="../ntvar">call ntvar</a> may be used as input to this program.</p>
<h3 id="mutation-database">Mutation Database</h3>
<p>The mutation database describes specific mutations within the named genetic regions specified previously in the BED4 file. When provided to the tool, the surveillence and drug resistance category information are included in the output mutation annotation.</p>
<p>The entries in the "genetic regions" colummn of this database must match the "names" column of the provided BED4 file. Please refer to <a href="../formats">Data Formats</a> for more information.</p>
<h2 id="options">Options</h2>
<h3 id="minimum-frequency">Minimum Frequency</h3>
<pre><code class="text">-f, --min_freq FLOAT
</code></pre>

<p>The minimum observed frequency for a variant to reported. The default frequency is 0.01.</p>
<h3 id="error-rate">Error Rate</h3>
<pre><code class="text">-e, --error_rate FLOAT
</code></pre>

<p>This is the expected substitution sequencing error rate. The default value is 0.0021 substitutions per sequenced base.</p>
<h3 id="output">Output</h3>
<pre><code class="text">-o, --output FILENAME
</code></pre>

<p>The file output location to write the identified amino acid mutations.</p>
<h2 id="output_1">Output</h2>
<p>The amino acid mutations that exceed the minimum frequency threshold will be output in <a href="https://github.com/winhiv/aavf-spec/blob/master/AAVFv1.0.pdf">AAVF</a> format. By default, the results will be printed to standard output. The user may direct the output to a file by specifying a file name with the <code>-o / --output</code> option.</p>
<h2 id="examples">Examples</h2>
<h3 id="data">Data</h3>
<p>The following example data may be used to run the tool:</p>
<ul>
<li><a href="../data/hiv.fasta">hiv.fasta</a></li>
<li><a href="../data/variant.bam">variant.bam</a></li>
<li><a href="../data/variant.bai">variant.bai</a></li>
<li><a href="../data/variant.vcf">variant.vcf</a></li>
<li><a href="../data/hiv_db.tsv">hiv_db.tsv</a></li>
</ul>
<h3 id="example-no-database">Example: No Database</h3>
<h4 id="command">Command</h4>
<pre><code>quasitools call aavar variant.bam hiv.fasta hiv.bed
</code></pre>

<h4 id="output_2">Output</h4>
<pre><code class="text">##reference=hiv.fasta
##source=quasitools:aavar
##fileformat=AAVFv1.0
##fileDate=20190206
##INFO=&lt;ID=SRVL,Number=.,Type=String,Description=&quot;Drug Resistance Surveillance&quot;&gt;
##INFO=&lt;ID=AC,Number=.,Type=String,Description=&quot;Alternate Codon&quot;&gt;
##INFO=&lt;ID=CAT,Number=.,Type=String,Description=&quot;Drug Resistance Category&quot;&gt;
##INFO=&lt;ID=ACF,Number=.,Type=Float,Description=&quot;Alternate Codon Frequency,for each Alternate Codon,in the same order aslisted.&quot;&gt;
##INFO=&lt;ID=RC,Number=1,Type=String,Description=&quot;Reference Codon&quot;&gt;
##FILTER=&lt;ID=af0.01,Description=&quot;Set if True; alt_freq&lt;0.01&quot;&gt;
#CHROM  GENE    POS REF ALT FILTER  ALT_FREQ    COVERAGE    INFO
AF033819.3  gag 339 P   Q   PASS    1.0000  133 SRVL=.;AC=cAa;CAT=.;ACF=1.0000;RC=cca
AF033819.3  env 67  N   T   PASS    1.0000  116 SRVL=.;AC=aCt;CAT=.;ACF=1.0000;RC=aat
AF033819.3  gag 441 Y   S   PASS    1.0000  109 SRVL=.;AC=tCc;CAT=.;ACF=1.0000;RC=tac
AF033819.3  pol 141 I   S   PASS    1.0000  145 SRVL=.;AC=aGt;CAT=.;ACF=1.0000;RC=att
AF033819.3  vpu 34  L   I   PASS    1.0000  118 SRVL=.;AC=Ata;CAT=.;ACF=1.0000;RC=tta
AF033819.3  env 302 N   Y   PASS    1.0000  138 SRVL=.;AC=Tat;CAT=.;ACF=1.0000;RC=aat
AF033819.3  gag 3   A   P   PASS    1.0000  140 SRVL=.;AC=Ccg;CAT=.;ACF=1.0000;RC=gcg
AF033819.3  pol 246 Q   H   PASS    1.0000  139 SRVL=.;AC=caT;CAT=.;ACF=1.0000;RC=caa
AF033819.3  gag 230 E   D   PASS    1.0000  126 SRVL=.;AC=gaT;CAT=.;ACF=1.0000;RC=gaa
AF033819.3  pol 96  G   E   PASS    1.0000  128 SRVL=.;AC=gAa;CAT=.;ACF=1.0000;RC=gga
</code></pre>

<p>Observe that there is no information under <code>INFO</code> column of the output for <code>SRVL</code> (surveillance) and <code>CAT</code> (drug resistance category). This is because a mutation database was not provided.</p>
<h3 id="example-with-database">Example: With Database</h3>
<h4 id="command_1">Command</h4>
<pre><code class="bash">quasitools call aavar variant.bam hiv.fasta hiv.bed variant.vcf hiv_db.tsv
</code></pre>

<h4 id="output_3">Output</h4>
<pre><code class="text">##reference=hiv.fasta
##source=quasitools:aavar
##fileformat=AAVFv1.0
##fileDate=20190206
##INFO=&lt;ID=SRVL,Number=.,Type=String,Description=&quot;Drug Resistance Surveillance&quot;&gt;
##INFO=&lt;ID=AC,Number=.,Type=String,Description=&quot;Alternate Codon&quot;&gt;
##INFO=&lt;ID=CAT,Number=.,Type=String,Description=&quot;Drug Resistance Category&quot;&gt;
##INFO=&lt;ID=ACF,Number=.,Type=Float,Description=&quot;Alternate Codon Frequency,for each Alternate Codon,in the same order aslisted.&quot;&gt;
##INFO=&lt;ID=RC,Number=1,Type=String,Description=&quot;Reference Codon&quot;&gt;
##FILTER=&lt;ID=af0.01,Description=&quot;Set if True; alt_freq&lt;0.01&quot;&gt;
#CHROM  GENE    POS REF ALT FILTER  ALT_FREQ    COVERAGE    INFO
AF033819.3  gag 339 P   Q   PASS    1.0000  133 SRVL=.;AC=cAa;CAT=.;ACF=1.0000;RC=cca
AF033819.3  env 67  N   T   PASS    1.0000  116 SRVL=No;AC=aCt;CAT=minor;ACF=1.0000;RC=aat
AF033819.3  gag 441 Y   S   PASS    1.0000  109 SRVL=No;AC=tCc;CAT=minor;ACF=1.0000;RC=tac
AF033819.3  pol 141 I   S   PASS    1.0000  145 SRVL=.;AC=aGt;CAT=.;ACF=1.0000;RC=att
AF033819.3  vpu 34  L   I   PASS    1.0000  118 SRVL=Yes;AC=Ata;CAT=major;ACF=1.0000;RC=tta
AF033819.3  env 302 N   Y   PASS    1.0000  138 SRVL=.;AC=Tat;CAT=.;ACF=1.0000;RC=aat
AF033819.3  gag 3   A   P   PASS    1.0000  140 SRVL=Yes;AC=Ccg;CAT=major;ACF=1.0000;RC=gcg
AF033819.3  pol 246 Q   H   PASS    1.0000  139 SRVL=No;AC=caT;CAT=minor;ACF=1.0000;RC=caa
AF033819.3  gag 230 E   D   PASS    1.0000  126 SRVL=.;AC=gaT;CAT=.;ACF=1.0000;RC=gaa
AF033819.3  pol 96  G   E   PASS    1.0000  128 SRVL=Yes;AC=gAa;CAT=major;ACF=1.0000;RC=gga
</code></pre>

<p>Observe that under the <code>INFO</code> column of the output, that the <code>SRVL</code> and <code>CAT</code> information is included in the annotation, if the mutation was specified in the mutation database.</p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../codonvar/" class="btn btn-neutral float-right" title="call codonvar">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../aacoverage/" class="btn btn-neutral" title="aacoverage"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
      <p>Copyright Government of Canada 2017-2019</p>
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../aacoverage/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../codonvar/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
